#### The principal choice of XGBoost model:

Linear regression was chosen as a simple benchmark to highlight the benefits of more powerful models like XGBoost, which can capture non-linear patterns and temporal effects that linear regression cannot.# We're gonna choose for benchmarking and interpretability linear regression, we aren't expecting to better  model than XGBoost because it struggle on handling non linearity patterns compared to tree based model, the reason of choosing just to understand regression task 

In [1]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet
from sklearn.metrics import  mean_squared_error, r2_score
import numpy as np

In [2]:
#Load datasets

train_day_df = pd.read_csv("D:/VS Code Projects/Datasets/Bike Sharing/data_processed/feature_engineered_train_day.csv")
valid_day_df = pd.read_csv("D:/VS Code Projects/Datasets/Bike Sharing/data_processed/feature_engineered_valid_day.csv")
train_hour_df = pd.read_csv("D:/VS Code Projects/Datasets/Bike Sharing/data_processed/feature_engineered_train_hour.csv")
valid_hour_df = pd.read_csv("D:/VS Code Projects/Datasets/Bike Sharing/data_processed/feature_engineered_valid_hour.csv")

In [3]:
### Ensure datetime & remove the column 'Unnamed: 0'
datasets_list =  [train_day_df,train_hour_df,valid_day_df,valid_hour_df]
for dataset in datasets_list:
    dataset['date'] = pd.to_datetime(dataset['date'], errors= 'coerce')
    dataset.drop(columns=['Unnamed: 0'], inplace=True, errors='ignore')

print('ISSUE FIXED')
train_day_df.dtypes

ISSUE FIXED


date                    datetime64[ns]
month                            int64
weather_situation                int64
feels_like_temp_norm           float64
temp_feel_diff                 float64
wind_speed                     float64
temp_x_wind_speed              float64
casual                           int64
registered                       int64
num_rentals                      int64
dtype: object

In [4]:
# ================================================
# Define target & features
# ================================================
target = 'num_rentals'
X_day_train = train_day_df.drop(columns= [target, 'date'])
y_day_train = train_day_df[target]

X_day_valid = valid_day_df.drop(columns= [target, 'date'])
y_day_valid =valid_day_df[target]

In [5]:

X_hour_train = train_hour_df.drop(columns= [target, 'date'])
y_hour_train = train_hour_df[target]

X_hour_valid = valid_hour_df.drop(columns= [target, 'date'])
y_hour_valid =valid_hour_df[target]

In [6]:
# ================================================
# Standardization (fit on train, transform eval)
# ================================================

scaler = StandardScaler()
X_day_train_scaled = scaler.fit_transform(X_day_train)
X_day_valid_scaled = scaler.fit_transform(X_day_valid)

X_hour_train_scaled = scaler.fit_transform(X_hour_train)
X_hour_valid_scaled = scaler.fit_transform(X_hour_valid)

In [7]:
# ================================================
# Train & Evaluate Models
# ================================================
lr =LinearRegression()


In [8]:
lr.fit(X_day_train_scaled, y_day_train)
y_day_pred = lr.predict(X_day_valid_scaled)

In [9]:
lr.fit(X_hour_train_scaled, y_hour_train)
y_hour_pred = lr.predict(X_hour_valid_scaled)

In [10]:
# Prediction Metrics
def evaluate (y_true, y_pred, label):
    rmse = np.sqrt(mean_squared_error(y_true, y_pred))
    r2 = r2_score(y_true, y_pred)
    print(f"{label} → RMSE: {rmse:.2f}, R²: {r2:.3f}")

In [11]:
evaluate(y_day_valid, y_day_pred, "Daily Linear Regression")

Daily Linear Regression → RMSE: 2373.10, R²: -0.915


In [12]:
evaluate(y_hour_valid, y_hour_pred, "Hourly Linear Regression")

Hourly Linear Regression → RMSE: 123.66, R²: 0.657


The model performs decently, slightly better than the persistence baseline (R² ~0.657, RMSE ~124), but not as well as XGBoost (R² 0.957, RMSE 43.86). It likely struggles because hourly rentals exhibit non-linear patterns, which XGBoost handles very well.  

For the daily dataset, the Linear Regression model performs even worse compared to XGBoost.
